File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.py
# 최소신장트리(크루스칼)

import sys
input = sys.stdin.readline

def find_parent(parent, x):
    if parent[x] != x:
        parent[x] = find_parent(parent, parent[x])
    return parent[x]

def union_parent(parent, a, b):
    a = find_parent(parent, a)
    b = find_parent(parent, b)
    if a < b:
        parent[b] = a
    else:
        parent[a] = b

V, E = map(int, input().split())

# 부모 테이블
parent = [0] * (V+1)
for i in range(1, V+1):
    parent[i] = i

edges = [] # 간선 정보
total_cost = 0 # 최소 신장 트리 계산

for _ in range(E):
    a, b, cost = map(int, input().split())
    edges.append((cost, a, b))

edges.sort()

for i in range(E):
    cost, a, b = edges[i]
    # find로 부모 노드 다르면 사이클 발생하지 않음
    # 사이클 발생하지 않으면 union으로 MST 추가
    if find_parent(parent, a) != find_parent(parent, b):
        union_parent(parent, a, b)
        total_cost += cost

print(total_cost)

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.java
import java.util.*;
import java.io.*;

public class Main {
    static int v, e;
    static ArrayList<ArrayList<Node>> graph;
    static int[] parent;

    static class Node implements Comparable<Node> {
        int to, from, cost;

        Node(int to, int from, int cost) {
            this.to = to;
            this.from = from;
            this.cost = cost;
        }

        @Override
        public int compareTo(Node e) {
            return this.cost - e.cost;
        }

    }

    static int find(int x) {
        if (parent[x] == x) return x;
        return parent[x] = find(parent[x]);
    }

    static void union(int x, int y) {
        x = find(x);
        y = find(y);
        if (x != y) parent[y] = x;
    }

    static boolean isSameParent(int x, int y) {
        x = find(x);
        y = find(y);
        if (x == y) return true;
        return false;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        v = Integer.parseInt(st.nextToken());
        e = Integer.parseInt(st.nextToken());

        parent = new int[v+1];
        for (int i = 1; i <= v; i++) parent[i] = i;

        Queue<Node> pq = new PriorityQueue<>();
        for (int i = 0; i < e; i++) {
            st = new StringTokenizer(br.readLine());
            int a, b, c;

            a = Integer.parseInt(st.nextToken());
            b = Integer.parseInt(st.nextToken());
            c = Integer.parseInt(st.nextToken());
            pq.offer(new Node(a, b, c));
        }

        int answer = 0;
        int size = pq.size();
        for (int i = 0; i < size; i++) {
            Node node = pq.poll();
            int to = find(node.to);
            int from = find(node.from);

            if (!isSameParent(to, from)) {
                answer += node.cost;
                union(node.to, node.from);
            }
        }

        System.out.println(answer);
    }
}